sb3_contrib/qrdqn/qrdqn.py

```diff
@@ -0,0 +1,250 @@
+        self.exploration_rate = 0.0
+        # Linear schedule will be defined in `_setup_model()`
+        self.exploration_schedule = None
+        self.quantile_net, self.quantile_net_target = None, None
+
+        if "optimizer_class" not in self.policy_kwargs:
+            self.policy_kwargs["optimizer_class"] = th.optim.Adam
+            self.policy_kwargs["optimizer_kwargs"] = dict(eps=0.01 / batch_size)  # proposed in the QRDQN paper
+
+        if _init_setup_model:
+            self._setup_model()
+
+    def _setup_model(self) -> None:
+        super(QRDQN, self)._setup_model()
+        self._create_aliases()
+        self.exploration_schedule = get_linear_fn(
+            self.exploration_initial_eps, self.exploration_final_eps, self.exploration_fraction
+        )
+
+    def _create_aliases(self) -> None:
+        self.quantile_net = self.policy.quantile_net
+        self.quantile_net_target = self.policy.quantile_net_target
+        self.n_quantiles = self.policy.n_quantiles
+
+    def _on_step(self) -> None:
+        """
+        Update the exploration rate and target network if needed.
+        This method is called in ``collect_rollouts()`` after each step in the environment.
+        """
+        if self.num_timesteps % self.target_update_interval == 0:
+            polyak_update(self.quantile_net.parameters(), self.quantile_net_target.parameters(), self.tau)
+
+        self.exploration_rate = self.exploration_schedule(self._current_progress_remaining)
+        logger.record("rollout/exploration rate", self.exploration_rate)
+
+    def train(self, gradient_steps: int, batch_size: int = 100) -> None:
+        # Update learning rate according to schedule
+        self._update_learning_rate(self.policy.optimizer)
+
+        losses = []
+        for gradient_step in range(gradient_steps):
+            # Sample replay buffer
+            replay_data = self.replay_buffer.sample(batch_size, env=self._vec_normalize_env)
+
+            with th.no_grad():
+                # Compute the quantiles of next observation
+                next_quantiles = self.quantile_net_target(replay_data.next_observations)
+                # Follow greedy policy: use the one with the highest value
+                next_quantiles, _ = next_quantiles.max(dim=2)
+                # 1-step TD target
+                target_quantiles = replay_data.rewards + (1 - replay_data.dones) * self.gamma * next_quantiles
+
+            # Get current quantile estimates
+            current_quantiles = self.quantile_net(replay_data.observations)
+
+            # Retrieve the quantiles for the actions from the replay buffer
+            actions = replay_data.actions[..., None].long().expand(batch_size, self.n_quantiles, 1)
+            current_quantiles = th.gather(current_quantiles, dim=2, index=actions).squeeze(2)
+
+            # Compute Quantile Huber loss
+            loss = quantile_huber_loss(current_quantiles, target_quantiles) * self.n_quantiles
```

Comment:
why do you multiply by `self.n_quantiles`? (that was not done in TQC original code if I recall... I did not check for QR-DQN yet)

Reply:
We sum over a quantile dimension, which is common in QR-DQN, IQN, and FQF.
I'm not sure why they take the mean in TQC.

Reply:
It seems that Dopamine is using the mean loss by default: https://github.com/google/dopamine/blob/master/dopamine/jax/agents/quantile/quantile_agent.py#L91
The same goes for IQN implementation of Pfrl: https://github.com/pfnet/pfrl/blob/master/pfrl/agents/iqn.py#L211

or the Facebook ReAgent implementation: https://github.com/facebookresearch/ReAgent/blob/master/reagent/training/qrdqn_trainer.py#L157
maybe that could be a parameter? (and check how it affects the learning)

Reply:
> It seems that Dopamine is using the mean loss by default: https://github.com/google/dopamine/blob/master/dopamine/jax/agents/quantile/quantile_agent.py#L91
The same goes for IQN implementation of Pfrl: https://github.com/pfnet/pfrl/blob/master/pfrl/agents/iqn.py#L211

Their implementations sum over the quantile dimension, so the same as mine, aren't they?
Multiplying "n_quantiles" means that summing over the quantile dimension.

Reply:
Maybe we should add an argument like `sum_over_quantiles : bool` to quantile_huber_loss?

Reply:
>Their implementations sum over the (current) quantile dimension, so the same as mine, aren't they?

I will try to check later when I'm rested ;)

>Maybe we should add an argument like sum_over_quantiles : bool to quantile_huber_loss?


Probably, yes